service/worker/scanner/executions/scavenger.go

```diff
@@ -0,0 +1,193 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package executions
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/cadence/client/frontend"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+	p "github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/service/worker/scanner/executor"
+)
+
+type ()
+
+type (
+	// Scavenger is the type that holds the state for executions scavenger daemon
+	Scavenger struct {
+		params ScannerWorkflowParams
+		// TODO: currently frontend client does not support scan visibility records without domain filter (need to chat with Bowei to understand if querying without domain filter is an issue)? Maybe we go directly to elasticSearch?
+		frontendClient frontend.Client // used to query visibility
+		historyDB      p.HistoryManager
+		executor       executor.Executor
+		metrics        metrics.Client
+		logger         log.Logger
+		stats          stats
+		status         int32
+		stopC          chan struct{}
+		stopWG         sync.WaitGroup
+	}
+
+	// ScannerWorkflowParams are the parameters passed to the executions scanner workflow
+	ScannerWorkflowParams struct {
+		VisibilityQuery string // optionally can be provided to limit the scope of the scan
+
+		// add other fields here such as: bool generateReport, string outputLocation, bool runInDryMode etc...
+	}
+
+	executionKey struct {
+		domainID   string
+		workflowID string
+		runID      string
+	}
+
+	stats struct {
+		// TODO: include stats here that should be tracked throughout execution of scavenger
+	}
+
+	// executorTask is a runnable task that adheres to the executor.Task interface
+	// for the scavenger, each of this task processes a single workflow execution
+	executorTask struct {
+		executionKey
+		scvg *Scavenger
+	}
+)
+
+var (
+	executionsBatchSize      = 32   // maximum number of executions we process concurrently
+	executionsPageSize       = 1000 // page size of executions read from visibility manager
+	executorPollInterval     = time.Minute
+	executorMaxDeferredTasks = 10000
+)
+
+// NewScavenger returns an instance of executions scavenger daemon
+// The Scavenger can be started by calling the Start() method on the
+// returned object. Calling the Start() method will result in one
```

Comment:
Would be nice if we are able to resume from where we left off previously. Not a blocker though

Reply:
I agree - I filed a ticket https://github.com/uber/cadence/issues/3064